splice_spark2/src/main/spark2.2/com/splicemachine/spark2/splicemachine/SplicemachineContext.scala

```diff
@@ -489,68 +552,269 @@ class SplicemachineContext(options: Map[String, String]) extends Serializable {
+    
+    var i = 0
+    var res = Seq.empty[String]
+    while( itr.hasNext ) {
+      val hashcode = externalizable(itr.next, schema, partition).hashCode(cols)
+      res = res :+ s"$i,${khashcodes.indexOf(hashcode)}\t${hashcode==lastKHash}"
+      i += 1
+    }
+    
+    println(s"$id SMC.checkRecovery res: Kafka count ${khashcodes.size}\n${res.mkString("\n")}")
+  }
+  
+  private[this] var sendDataTimestamp: Long = _
+  
+  private[this] def sendData(topicName: String, rdd: JavaRDD[Row], schema: StructType): Unit = {
+    sendDataTimestamp = System.currentTimeMillis
     rdd.rdd.mapPartitionsWithIndex(
       (partition, itrRow) => {
-        val taskContext = TaskContext.get
+        val id = topicName.substring(0,5)+":"+partition.toString
+        //println(s"$id SMC.sendData p== $partition")
 
-        var msgIdx = 0
-        if (taskContext != null && taskContext.attemptNumber > 0) {
-          val entriesInKafka = KafkaUtils.messageCount(kafkaServers, topicName, partition)
-          for(i <- (1: Long) to entriesInKafka) {
-            itrRow.next
+        var msgCount = 0
+        val taskContext = TaskContext.get
+        val itr = if (taskContext != null && taskContext.attemptNumber > 0) {
+          // Recover from previous task failure
+          // Be sure the iterator is advanced past the items previously published to Kafka
+          
+          println(s"$id SMC.sendData Retry $partition ${taskContext.attemptNumber} ${insAccum.sum}")
+
+          //val itr12 = itrRow //.duplicate
+          //checkRecovery(id, topicName, partition, itr12._1, schema)
+
+          val lastMsg = KafkaUtils.lastMessageOf(kafkaServers, topicName, partition)
+          if( lastMsg.isEmpty ) {
+            itrRow
+          } else {
+            val lastVR = lastMsg.get.asInstanceOf[KafkaReadFunction.Message].vr
+            val hashCols = if( lastVR.length > 0) { Range(0,lastVR.length-1).toArray } else { Array(0) }
+            val lastKHash = lastVR.hashCode(hashCols)
+            def hash: Row => Int = row => externalizable(row, schema, partition).hashCode(hashCols)
+            
+            //val itr34 = itr12._2.duplicate
+            val itr34 = itrRow.duplicate   //itr12.duplicate
+            
+            val inKafka_NotInKafka = itr34._1.span( hash(_) != lastKHash )
+            if( inKafka_NotInKafka._2.hasNext ) {
+              inKafka_NotInKafka._2.next  // matches the last item in Kafka, get past it
+              //println(s"$id SMC.sendData Retry skip ${hash(r)} $lastKHash")
+              msgCount = inKafka_NotInKafka._1.size + 1  // this message count was lost during previous task failure
+              inKafka_NotInKafka._2
+            } else {  // itrRow starts after the last item added to Kafka
+              itr34._2
+            }
           }
-          msgIdx = entriesInKafka.asInstanceOf[Int]
+        } else {
+          itrRow
```

Comment:
Can you explain a bit what's going on here? The naming of variables is a bit confusing.

Reply:
Added comments to the code in a temporary branch for review while regression tests are running, and will push to this branch after regression is finished.

https://github.com/splicemachine/spliceengine/blob/8e1479e588459de7fb11b65b7529890a3c71a7de/splice_spark2/src/main/spark2.4/com/splicemachine/spark2/splicemachine/SplicemachineContext.scala#L738-L768

Reply:
Pushed code to this branch.